polars can't read `timestamp[s]` typed columns in parquet files made by pyarrow

#### Are you using Python or Rust?

**Python**

#### What version of polars are you using?

polars-0.12.20

#### What operating system are you using polars on?

Linux (Debian 11)

#### Describe your bug.

Polars cannot accurately read the datetime from Parquet files created with `timestamp[s]` in pyarrow.

```python
>>> pl.read_parquet("test.parquet")
shape: (2, 2)
┌───────────────────────┬─────────────────────┐
│ datetime[s]           ┆ datetime[ms]        │
│ ---                   ┆ ---                 │
│ datetime[ms]          ┆ datetime[ms]        │
╞═══════════════════════╪═════════════════════╡
│ +52671-12-25 12:26:40 ┆ 2020-09-13 12:26:40 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ +55840-11-08 22:13:20 ┆ 2023-11-14 22:13:20 │
└───────────────────────┴─────────────────────┘
```

I have not been able to determine if this is a problem with polars or arrow2, but since multiple Parquet readers other than polars did not have the problem, I do not think it is a problem with the official arrow library. I apologize if this is not a submission to the appropriate repository.

#### What are the steps to reproduce the behavior?

```python
import polars as pl
import pyarrow as pa
import pyarrow.parquet as pq

pt = pa.table([
    pa.array([1600000000, 1700000000], type=pa.timestamp("s")),
    pa.array([1600000000000, 1700000000000], type=pa.timestamp("ms"))
], names=["datetime[s]", "datetime[ms]"])

pq.write_table(pt, "test.parquet")

pl.read_parquet("test.parquet")
```

#### What is the expected behavior?

It needs to be interpreted as exact time like other tools such as pyarrow.

```python
>>> pl.from_arrow(pq.read_table("test.parquet"))
shape: (2, 2)
┌─────────────────────┬─────────────────────┐
│ datetime[s]         ┆ datetime[ms]        │
│ ---                 ┆ ---                 │
│ datetime[ms]        ┆ datetime[ms]        │
╞═════════════════════╪═════════════════════╡
│ 2020-09-13 12:26:40 ┆ 2020-09-13 12:26:40 │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2023-11-14 22:13:20 ┆ 2023-11-14 22:13:20 │
└─────────────────────┴─────────────────────┘
```

@jorgecarleitao is this maybe a conversion error in arrow2/parquet2?

```python
import pyarrow.parquet as pq
seconds = [1600000000, 1700000000]

pt = pa.table([
    pa.array(seconds, type=pa.timestamp("s")),
], names=["datetime[s]"])

pq.write_table(pt, "test.parquet")
df = pl.read_parquet("test.parquet")

# undo the conversion done by polars
seconds_read = df.to_series().cast(int) // 1000

for a, b in zip(seconds_read, seconds):
    print(a - b)

assert seconds_read.to_list() == seconds
```

```
1598400000000
1698300000000

---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
/tmp/ipykernel_58298/418428865.py in <module>
     15     print(a - b)
     16 
---> 17 assert seconds_read.to_list() == seconds

AssertionError: 


```


## Writing /reading arrow2

Writing and reading timestamp with `ms` unit seems to go fine.

```python
seconds = [1600000000, 1700000000]
df = pl.DataFrame({
    "time": seconds
}).with_column(pl.col("time").cast(pl.Datetime))

df.to_parquet("test.parquet")
df = pl.read_parquet("test.parquet")

assert df.to_series().cast(int).to_list() == seconds
```

--- a/polars/polars-arrow/Cargo.toml
+++ b/polars/polars-arrow/Cargo.toml
@@ -9,8 +9,8 @@ description = "Arrow interfaces for Polars DataFrame library"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-# arrow = { package = "arrow2", git = "https://github.com/jorgecarleitao/arrow2", rev = "b46a636f31d70a20bd54df9c7f9e9363053ade08", default-features = false }
-arrow = { package = "arrow2", git = "https://github.com/ritchie46/arrow2", branch = "cherry_pick", default-features = false }
+arrow = { package = "arrow2", git = "https://github.com/jorgecarleitao/arrow2", rev = "0175d9e3c54f6c3c81e3a6db3f1c39b4ba3b93fe", default-features = false }
+# arrow = { package = "arrow2", git = "https://github.com/ritchie46/arrow2", branch = "cherry_pick", default-features = false }
 # arrow = { package = "arrow2", version = "0.9", default-features = false, features = ["compute_concatenate"] }
 hashbrown = "0.12"
 num = "^0.4"

--- a/polars/polars-core/Cargo.toml
+++ b/polars/polars-core/Cargo.toml
@@ -160,10 +160,10 @@ unsafe_unwrap = "^0.1.0"
 
 [dependencies.arrow]
 package = "arrow2"
-# git = "https://github.com/jorgecarleitao/arrow2"
-git = "https://github.com/ritchie46/arrow2"
-# rev = "b46a636f31d70a20bd54df9c7f9e9363053ade08"
-branch = "cherry_pick"
+git = "https://github.com/jorgecarleitao/arrow2"
+# git = "https://github.com/ritchie46/arrow2"
+rev = "0175d9e3c54f6c3c81e3a6db3f1c39b4ba3b93fe"
+# branch = "cherry_pick"
 # version = "0.9"
 default-features = false
 features = [

--- a/polars/polars-io/Cargo.toml
+++ b/polars/polars-io/Cargo.toml
@@ -16,7 +16,7 @@ json = ["arrow/io_json"]
 ipc = ["arrow/io_ipc", "arrow/io_ipc_compression"]
 # ipc = []
 lazy = []
-parquet = ["polars-core/parquet", "arrow/io_parquet", "arrow/io_parquet_compression", "polars-utils"]
+parquet = ["polars-core/parquet", "arrow/io_parquet", "arrow/io_parquet_compression", "polars-utils", "memmap"]
 dtype-datetime = ["polars-core/dtype-datetime", "polars-core/temporal"]
 dtype-date = ["polars-core/dtype-date"]
 dtype-time = ["polars-core/dtype-time", "polars-core/temporal"]
@@ -31,8 +31,8 @@ private = []
 [dependencies]
 ahash = "0.7"
 anyhow = "1.0"
-# arrow = { package = "arrow2", git = "https://github.com/jorgecarleitao/arrow2", rev = "b46a636f31d70a20bd54df9c7f9e9363053ade08", default-features = false }
-arrow = { package = "arrow2", git = "https://github.com/ritchie46/arrow2", branch = "cherry_pick", default-features = false }
+arrow = { package = "arrow2", git = "https://github.com/jorgecarleitao/arrow2", rev = "0175d9e3c54f6c3c81e3a6db3f1c39b4ba3b93fe", default-features = false }
+# arrow = { package = "arrow2", git = "https://github.com/ritchie46/arrow2", branch = "cherry_pick", default-features = false }
 # arrow = { package = "arrow2", version = "0.9", default-features = false }
 csv-core = { version = "0.1.10", optional = true }
 dirs = "4.0"

--- a/polars/polars-io/src/parquet/predicates.rs
+++ b/polars/polars-io/src/parquet/predicates.rs
@@ -107,11 +107,8 @@ pub(crate) fn collect_statistics(
     let mut fields = vec![];
     let mut stats = vec![];
 
-    for (column_chunk_md, fld) in md.iter().zip(&schema.fields) {
-        if let Some(parquet_stats) = column_chunk_md.statistics() {
-            let parquet_stats = parquet_stats?;
-            let st = deserialize_statistics(&*parquet_stats)?;
-
+    for fld in &schema.fields {
+        for st in deserialize_statistics(fld, md)?.into_iter().flatten() {
             fields.push(fld.into());
             stats.push(ColumnStats(st));
         }

--- a/polars/polars-io/src/parquet/read.rs
+++ b/polars/polars-io/src/parquet/read.rs
@@ -1,11 +1,9 @@
-use super::{ArrowReader, ArrowResult};
 use crate::aggregations::ScanAggregation;
 use crate::mmap::MmapBytesReader;
-use crate::parquet::read_impl::{parallel_read, read_parquet};
+use crate::parquet::read_impl::read_parquet;
 use crate::predicates::PhysicalIoExpr;
 use crate::prelude::*;
 use arrow::io::parquet::read;
-use polars_core::frame::ArrowChunk;
 use polars_core::prelude::*;
 use std::io::{Read, Seek};
 use std::sync::Arc;
@@ -32,15 +30,10 @@ impl<R: MmapBytesReader> ParquetReader<R> {
     ) -> Result<DataFrame> {
         // this path takes predicates and parallelism into account
         let metadata = read::read_metadata(&mut self.reader)?;
-        let schema = read::schema::get_schema(&metadata)?;
-
-        let f = match self.parallel {
-            true => parallel_read,
-            false => read_parquet,
-        };
+        let schema = read::schema::infer_schema(&metadata)?;
 
         let rechunk = self.rechunk;
-        f(
+        read_parquet(
             self.reader,
             self.n_rows.unwrap_or(usize::MAX),
             projection,
@@ -48,6 +41,7 @@ impl<R: MmapBytesReader> ParquetReader<R> {
             Some(metadata),
             predicate,
             aggregate,
+            self.parallel,
         )
         .map(|mut df| {
             if rechunk {
@@ -86,14 +80,8 @@ impl<R: MmapBytesReader> ParquetReader<R> {
     pub fn schema(mut self) -> Result<Schema> {
         let metadata = read::read_metadata(&mut self.reader)?;
 
-        let schema = read::get_schema(&metadata)?;
+        let schema = read::infer_schema(&metadata)?;
         Ok(schema.into())
-    }
-}
-
-impl<R: Read + Seek> ArrowReader for read::RecordReader<R> {
-    fn next_record_batch(&mut self) -> ArrowResult<Option<ArrowChunk>> {
-        self.next().map_or(Ok(None), |v| v.map(Some))
     }
 }
 
@@ -116,7 +104,7 @@ impl<R: MmapBytesReader> SerReader<R> for ParquetReader<R> {
 
     fn finish(mut self) -> Result<DataFrame> {
         let metadata = read::read_metadata(&mut self.reader)?;
-        let schema = read::schema::get_schema(&metadata)?;
+        let schema = read::schema::infer_schema(&metadata)?;
 
         if let Some(cols) = self.columns {
             let mut prj = Vec::with_capacity(cols.len());
@@ -128,26 +116,7 @@ impl<R: MmapBytesReader> SerReader<R> for ParquetReader<R> {
             self.projection = Some(prj);
         }
 
-        if self.parallel {
-            let rechunk = self.rechunk;
-            return parallel_read(
-                self.reader,
-                self.n_rows.unwrap_or(usize::MAX),
-                self.projection.as_deref(),
-                &schema,
-                Some(metadata),
-                None,
-                None,
-            )
-            .map(|mut df| {
-                if rechunk {
-                    df.rechunk();
-                };
-                df
-            });
-        }
-
-        let mut df = read_parquet(
+        read_parquet(
             self.reader,
             self.n_rows.unwrap_or(usize::MAX),
             self.projection.as_deref(),
@@ -155,11 +124,13 @@ impl<R: MmapBytesReader> SerReader<R> for ParquetReader<R> {
             Some(metadata),
             None,
             None,
-        )?;
-        if self.rechunk {
-            df.rechunk();
-        }
-
-        Ok(df)
+            self.parallel,
+        )
+        .map(|mut df| {
+            if self.rechunk {
+                df.rechunk();
+            }
+            df
+        })
     }
 }

--- a/polars/polars-io/src/parquet/read_impl.rs
+++ b/polars/polars-io/src/parquet/read_impl.rs
@@ -3,13 +3,11 @@ use crate::mmap::{MmapBytesReader, ReaderBytes};
 use crate::parquet::predicates::collect_statistics;
 use crate::predicates::{apply_predicate, arrow_schema_to_empty_df, PhysicalIoExpr};
 use crate::utils::apply_projection;
-use arrow::array::ArrayRef;
 use arrow::io::parquet::read;
-use arrow::io::parquet::read::{FileMetaData, MutStreamingIterator};
+use arrow::io::parquet::read::{to_deserializer, FileMetaData};
 use polars_core::prelude::*;
 use polars_core::utils::accumulate_dataframes_vertical;
 use polars_core::POOL;
-use polars_utils::contention_pool::LowContentionPool;
 use rayon::prelude::*;
 use std::borrow::Cow;
 use std::convert::TryFrom;
@@ -17,6 +15,7 @@ use std::io::Cursor;
 use std::ops::Deref;
 use std::sync::Arc;
 
+#[allow(clippy::too_many_arguments)]
 pub fn read_parquet<R: MmapBytesReader>(
     reader: R,
     limit: usize,
@@ -25,6 +24,7 @@ pub fn read_parquet<R: MmapBytesReader>(
     metadata: Option<FileMetaData>,
     predicate: Option<Arc<dyn PhysicalIoExpr>>,
     aggregate: Option<&[ScanAggregation]>,
+    parallel: bool,
 ) -> Result<DataFrame> {
     let reader = ReaderBytes::from(&reader);
     let bytes = reader.deref();
@@ -40,9 +40,6 @@ pub fn read_parquet<R: MmapBytesReader>(
         .unwrap_or_else(|| Cow::Owned((0usize..schema.fields.len()).collect::<Vec<_>>()));
 
     let mut dfs = Vec::with_capacity(row_group_len);
-
-    let mut buf_1 = Vec::with_capacity(1024);
-    let mut buf_2 = Vec::with_capacity(1024);
 
     let mut remaining_rows = limit;
 
@@ -61,38 +58,46 @@ pub fn read_parquet<R: MmapBytesReader>(
                 }
             }
         }
+
         // test we don't read the parquet file if this env var is set
         #[cfg(debug_assertions)]
         {
             assert!(std::env::var("POLARS_PANIC_IF_PARQUET_PARSED").is_err())
         }
 
-        let columns = projection
-            .clone()
-            .iter()
-            .map(|column_i| {
-                let b1 = std::mem::take(&mut buf_1);
-                let b2 = std::mem::take(&mut buf_2);
+        let chunk_size = md.num_rows() as usize;
+        let columns = if parallel {
+            POOL.install(|| {
+                projection
+                    .par_iter()
+                    .map(|column_i| {
+                        let mut reader = Cursor::new(bytes);
+                        let field = &schema.fields[*column_i];
+                        let columns = read::read_columns(&mut reader, md.columns(), &field.name)?;
+                        let mut iter = to_deserializer(
+                            columns,
+                            field.clone(),
+                            remaining_rows,
+                            Some(chunk_size),
+                        )?;
 
-                // the get_column_iterator is an iterator of columns, each column contains compressed pages.
-                // get_column_iterator yields `Vec<Vec<CompressedPage>>`:
-                // outer `Vec` is len 1 for primitive types,
-                // inner `Vec` is whatever number of pages the chunk contains.
-                let column_iter =
-                    read::get_column_iterator(&mut reader, &file_metadata, rg, *column_i, None, b1);
-                let fld = &schema.fields[*column_i];
-                let (mut array, b1, b2) = read::column_iter_to_array(column_iter, fld, b2)?;
+                        Series::try_from((field.name.as_str(), iter.next().unwrap()?))
+                    })
+                    .collect::<Result<Vec<_>>>()
+            })?
+        } else {
+            projection
+                .iter()
+                .map(|column_i| {
+                    let field = &schema.fields[*column_i];
+                    let columns = read::read_columns(&mut reader, md.columns(), &field.name)?;
+                    let mut iter =
+                        to_deserializer(columns, field.clone(), remaining_rows, Some(chunk_size))?;
 
-                if array.len() > remaining_rows {
-                    array = array.slice(0, remaining_rows);
-                }
-
-                buf_1 = b1;
-                buf_2 = b2;
-
-                Series::try_from((fld.name.as_str(), Arc::from(array)))
-            })
-            .collect::<Result<Vec<_>>>()?;
+                    Series::try_from((field.name.as_str(), iter.next().unwrap()?))
+                })
+                .collect::<Result<Vec<_>>>()?
+        };
 
         remaining_rows = file_metadata.row_groups[rg].num_rows() as usize;
 
@@ -109,133 +114,6 @@ pub fn read_parquet<R: MmapBytesReader>(
             Cow::Owned(apply_projection(schema, &projection))
         } else {
             Cow::Borrowed(schema)
-        };
-        Ok(arrow_schema_to_empty_df(&schema))
-    } else {
-        let mut df = accumulate_dataframes_vertical(dfs.into_iter())?;
-        apply_aggregations(&mut df, aggregate)?;
-        Ok(df.slice(0, limit))
-    }
-}
-
-pub(crate) fn parallel_read<R: MmapBytesReader>(
-    reader: R,
-    limit: usize,
-    projection: Option<&[usize]>,
-    arrow_schema: &ArrowSchema,
-    metadata: Option<FileMetaData>,
-    predicate: Option<Arc<dyn PhysicalIoExpr>>,
-    aggregate: Option<&[ScanAggregation]>,
-) -> Result<DataFrame> {
-    let reader = ReaderBytes::from(&reader);
-    let bytes = reader.deref();
-    let mut reader = Cursor::new(bytes);
-
-    let file_metadata = metadata
-        .map(Ok)
-        .unwrap_or_else(|| read::read_metadata(&mut reader))?;
-
-    let parq_fields = if let Some(projection) = projection {
-        let parq_fields = file_metadata.schema().fields();
-        Cow::Owned(
-            projection
-                .iter()
-                .map(|i| parq_fields[*i].clone())
-                .collect::<Vec<_>>(),
-        )
-    } else {
-        Cow::Borrowed(file_metadata.schema().fields())
-    };
-
-    let n_groups = file_metadata.row_groups.len();
-    let mut dfs = Vec::with_capacity(n_groups);
-
-    // we need to store two buffers to be reused, so the contention pool size
-    // is the number of threads * 3 b1, b2, and column_chunks
-    let pool_size = POOL.current_num_threads() * 2 + 1;
-
-    let cont_pool = LowContentionPool::<Vec<u8>>::new(pool_size);
-
-    for row_group in 0..n_groups {
-        let md = &file_metadata.row_groups[row_group];
-        if let Some(pred) = &predicate {
-            if let Some(pred) = pred.as_stats_evaluator() {
-                if let Some(stats) = collect_statistics(md.columns(), arrow_schema)? {
-                    let should_read = pred.should_read(&stats);
-                    // a parquet file may not have statistics of all columns
-                    if matches!(should_read, Ok(false)) {
-                        continue;
-                    } else if !matches!(should_read, Err(PolarsError::NotFound(_))) {
-                        let _ = should_read?;
-                    }
-                }
-            }
-        }
-        // test we don't read the parquet file if this env var is set
-        #[cfg(debug_assertions)]
-        {
-            assert!(std::env::var("POLARS_PANIC_IF_PARQUET_PARSED").is_err())
-        }
-
-        let columns = POOL.install(|| {
-            parq_fields
-                .par_iter()
-                .enumerate()
-                .map(|(mut field_i, field)| {
-                    if let Some(projection) = projection {
-                        field_i = projection[field_i];
-                    }
-
-                    // <IO bounded>
-
-                    // create a new reader
-                    let reader = Cursor::new(bytes);
-
-                    let b1 = cont_pool.get();
-                    // get compressed column pages
-                    let mut columns = read::get_column_iterator(
-                        reader,
-                        &file_metadata,
-                        row_group,
-                        field_i,
-                        None,
-                        b1,
-                    );
-
-                    let mut column_chunks = Vec::with_capacity(64);
-                    while let read::State::Some(mut new_iter) = columns.advance().unwrap() {
-                        if let Some((pages, metadata)) = new_iter.get() {
-                            let pages = pages.collect::<Vec<_>>();
-
-                            column_chunks.push((pages, metadata.clone()));
-                        }
-                        columns = new_iter;
-                    }
-
-                    // <CPU bounded>
-                    let columns = read::ReadColumnIterator::new(field.clone(), column_chunks);
-                    let field = &arrow_schema.fields[field_i];
-
-                    let b2 = cont_pool.get();
-                    let (arr, b1, b2) = read::column_iter_to_array(columns, field, b2)?;
-                    cont_pool.set(b1);
-                    cont_pool.set(b2);
-                    Series::try_from((field.name.as_str(), Arc::from(arr) as ArrayRef))
-                })
-                .collect::<Result<Vec<_>>>()
-        })?;
-        let mut df = DataFrame::new_no_checks(columns);
-        apply_predicate(&mut df, predicate.as_deref())?;
-        apply_aggregations(&mut df, aggregate)?;
-
-        dfs.push(df)
-    }
-
-    if dfs.is_empty() {
-        let schema = if let Some(proj) = projection {
-            Cow::Owned(apply_projection(arrow_schema, proj))
-        } else {
-            Cow::Borrowed(arrow_schema)
         };
         Ok(arrow_schema_to_empty_df(&schema))
     } else {

